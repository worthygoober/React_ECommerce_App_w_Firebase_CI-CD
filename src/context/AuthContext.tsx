import { createContext, useState, useEffect, useContext } from "react";
import { onAuthStateChanged, User } from "firebase/auth";
import { auth } from '../lib/firebase/firebase';

interface AuthState {
    user: User | null,
    setUser: (user: User) => void;
}

const AuthContext = createContext<AuthState>({
    user: null,
    setUser: (user: User) => {}
});

export const AuthProvider = ({ children }: { children: React.ReactNode}) => {
    const [user, setUser] = useState< User | null> (null);
    useEffect(() => {
        const unsubscribe = onAuthStateChanged(auth, (user) => {
            if (user) {
                setUser(user);
            } else {
                setUser(null);
            }
        });
        return () => unsubscribe();
    }, []);

    return (
        <AuthContext.Provider value={{ user, setUser }}>
            {children}
        </AuthContext.Provider>
    )
};

export const useAuth = () => useContext(AuthContext);

export default AuthContext;