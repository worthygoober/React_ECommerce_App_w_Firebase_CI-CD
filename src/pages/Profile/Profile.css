:root {
    font-family: system-ui, Avenir, Helvetica, Arial, sans-serif;
    line-height: 1.5;
    font-weight: 400;
  
    color-scheme: light dark;
    color: rgba(255, 255, 255, 0.87);
    background-color: #242424;
  
    font-synthesis: none;
    text-rendering: optimizeLegibility;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
  }

.form {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    height: 100vh;
}

.fieldset {
    border: 1px solid #ccc;
    padding: 20px;
    border-radius: 5px;
    width: 300px;
    margin: 0 auto;
}

.legend {
    font-size: 20px;
    font-weight: bold;
    margin-bottom: 10px;
}

.input {
    display: block;
    margin-bottom: 10px;
    padding: 10px 5px;
    border-radius: 5px;
    border: 1px solid #ccc;
    width: 100%;
}

.error {
    color: red;
    margin-bottom: 10px;
}

.success {
    color: green;
    margin-bottom: 10px;
}

.button {
    background-color: blue;
    color: white;
    border: none;
    padding: 10px;
    border-radius: 5px;
    cursor: pointer;
    transition: background 0.2s;
    display: block;
    margin: 20px auto;
    text-decoration: none;
    width: 25%;
    font-size: 0.9rem;
    text-align: center;
}

.button:hover {
    background: #3367d6;
  }

.deleteButton {
    background-color: red;
    color: white;
    border: none;
    padding: 10px;
    border-radius: 5px;
    cursor: pointer;
    transition: background 0.2s;
    display: block;
}

.deleteButton:hover {
    background: #ff6b6b;
  }