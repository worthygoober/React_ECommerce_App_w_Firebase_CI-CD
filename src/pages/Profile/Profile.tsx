import { useEffect, useState } from "react";
import { useAuth } from "../../context/AuthContext";
import { updateProfile, deleteUser } from "firebase/auth";
import { doc, getDoc, updateDoc, deleteDoc, collection, query, where, getDocs, Timestamp } from "firebase/firestore";
import { auth, db } from "../../lib/firebase/firebase";
import { onAuthStateChanged } from "firebase/auth";
import { useNavigate } from "react-router-dom";
import './Profile.css';
import { Order } from "../../types/types";
import { Flip, ToastContainer, toast } from "react-toastify";

const Profile = () => {
  const { user } = useAuth();
  const [displayName, setDisplayName] = useState(user?.displayName || '');
  const [email] = useState(user?.email || '');
  const [address, setAddress] = useState('');
  const [orders, setOrders] = useState<Order[]>([]);
  const [expandedOrder, setExpandedOrder] =  useState<string | null>(null);

  const navigate = useNavigate();

  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, (user) => {
      if (!user) {
        navigate('/login');
      }
    });
    return () => unsubscribe();
  }, [navigate]);

  useEffect(() => {
    console.log('User state in useEffect:', user);
    const loadUserData = async () => {
      if (!user?.uid) {
        console.log('User UID is missing, exiting function');
        return;
      }

      try {
        const userDoc = await getDoc(doc(db, 'users', user.uid));
        if (userDoc.exists()) {
          const userData = userDoc.data();
          console.log("Fetched user data from Firestore:", userData);
          setDisplayName(user?.displayName || userData.displayName || '');
          setAddress( userData.address || '');
        } else {
          console.log('User document does not exist.');
        }
      } catch (err: any) {
        toast.error(`Failed to load profile data: ${err.message}`, {
          autoClose: 5000,
        });
        console.error('Error fetching user data:', err);
      }
    };
    loadUserData();
  }, [user]);

  useEffect(() => {
    if (!user) return;

    const fetchOrders = async () => {
      try {
        const orderQuery = query(collection(db, 'orders'), where('userId', '==', user.uid));
        const querySnapshot = await getDocs(orderQuery);

        const fetchedOrders: Order[] = querySnapshot.docs.map((doc) => {
          const data = doc.data() as Omit<Order, "id">;
          return { id: doc.id, ...data };
        });

        setOrders(fetchedOrders);
      } catch (error: any) {
        toast.error(`Failed to fetch order data: ${error.message}`, {
          autoClose: 5000,
        });
        console.error("Error fetching orders:", error);
      }
    };

    fetchOrders();
  }, [user]);

  const handleUpdateProfile = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    if (!user) {
      toast.error('User not found');
      return;
    }

    try {
      await updateProfile(user, {
        displayName: displayName,
      });

      await updateDoc(doc(db, 'users', user.uid), {
        address: address,
        displayName: displayName,
        updatedAt: new Date()
      });

      toast.success('Profile updated successfully');
    } catch (error: any) {
      toast.error(`Profile update failed!: ${error.message}`, {
        autoClose: 5000,
      });
      console.error('Error updating profile:', error);
    }
  };

  const handleDeleteAccount = async () => {
    if (!user?.uid) return;

    try{
      if (!user) {
        toast.error('User not found');
        return;
      }

      await deleteDoc(doc(db, 'users', user.uid));

      await deleteUser(user);

      toast.success('Account deleted successfully');
      navigate('/register');
    } catch (error: any) {
      toast.error(`Profile deletion failed!: ${error.message}`, {
        autoClose: 5000,
      });
      console.error('Error deleting profile:', error);
    }
  };
  
  void auth;

  return (
    <div className="form">
      <h1>Profile</h1>

      <h3>Welcome back, {user?.displayName}</h3>

      <form onSubmit={handleUpdateProfile}>
        <fieldset className="fieldset">
        <legend className="legend">Profile</legend>

          <input
          className="input"
          type="text"
          value={displayName}
          placeholder="Name"
          onChange={(e) => setDisplayName(e.target.value)}
          />

          <input
          className="input"
          type="text"
          value={address}
          placeholder="Address"
          onChange={(e) => setAddress(e.target.value)}
          />

          <input
          className="input"
          type="email"
          value={email}
          placeholder="Email"
          disabled={true}
          />

        </fieldset>
        <button type="submit" className="button">
          Update Profile
        </button>
      </form>
      <button className="deleteButton" onClick={handleDeleteAccount}>
            Delete Account
      </button>

      <div>
        <h2>Order History</h2>
        {orders.length === 0 ? 
        <p>No orders found.</p> : (
          <ul>
            {orders.map(order => (
              <li key={order.id}>
                <button onClick={() => setExpandedOrder(expandedOrder === order.id ? null : order.id)}>
                  Order ID: {order.id} (Total: ${order.totalPrice.toFixed(2)})
                </button>
                {expandedOrder === order.id && (
                  <div>
                    <p>Placed on: {order.createdAt instanceof Timestamp ? order.createdAt.toDate().toLocaleString() : 'Unknown Date'}</p>
                    <ul>
                      {order.items.map(item => (
                        <li key={item.id}>
                          {item.title} - {item.quantity} x ${item.price}
                        </li>
                      ))}
                    </ul>
                  </div>
                )}
              </li>
            ))}
          </ul>
        )}
      </div>
      <ToastContainer 
        position="top-center"
        autoClose={3000}
        hideProgressBar={false}
        newestOnTop={false}
        closeOnClick={false}
        rtl={false}
        pauseOnFocusLoss
        draggable
        pauseOnHover
        theme="dark"
        transition={Flip}
      />
    </div>
  );
};

export default Profile;