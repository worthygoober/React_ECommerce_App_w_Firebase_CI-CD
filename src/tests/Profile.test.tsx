// Snapshot testing

import { render, waitFor } from "@testing-library/react";
import Profile from "../pages/Profile/Profile";
import { MemoryRouter, useNavigate } from "react-router-dom"; // Wrap in Router for useNavigate to work

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(), // Mock useNavigate
}));

jest.mock('../context/AuthContext', () => ({
  useAuth: jest.fn(),
}));

import { useAuth } from "../context/AuthContext";

test('matches snapshot for Profile', async () => {
  const mockedNavigate = jest.fn(); // Mock navigate function
  (useNavigate as jest.Mock).mockReturnValue(mockedNavigate); // Mock the navigate return value

  (useAuth as jest.Mock).mockReturnValue({ user: null });

  // Render the Profile component with MemoryRouter
  const { asFragment } = render(
    <MemoryRouter>
        <Profile />
    </MemoryRouter>
  );

  // Create a snapshot of the rendered component
  expect(asFragment()).toMatchSnapshot();

  await waitFor(() => {
    expect(mockedNavigate).toHaveBeenCalledWith('/login');
  });
});