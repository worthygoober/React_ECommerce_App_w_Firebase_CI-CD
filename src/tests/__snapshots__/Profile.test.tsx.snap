// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot for Profile 1`] = `
<DocumentFragment>
  <div
    class="form"
  >
    <h1>
      Profile
    </h1>
    <h3>
      Welcome back, 
    </h3>
    <form>
      <fieldset
        class="fieldset"
      >
        <legend
          class="legend"
        >
          Profile
        </legend>
        <input
          class="input"
          placeholder="Name"
          type="text"
          value=""
        />
        <input
          class="input"
          placeholder="Address"
          type="text"
          value=""
        />
        <input
          class="input"
          disabled=""
          placeholder="Email"
          type="email"
          value=""
        />
      </fieldset>
      <button
        class="button"
        type="submit"
      >
        Update Profile
      </button>
    </form>
    <button
      class="deleteButton"
    >
      Delete Account
    </button>
    <div>
      <h2>
        Order History
      </h2>
      <p>
        No orders found.
      </p>
    </div>
    <section
      aria-atomic="false"
      aria-label="Notifications Alt+T"
      aria-live="polite"
      aria-relevant="additions text"
      class="Toastify"
    />
  </div>
</DocumentFragment>
`;
